File: verso/lib/shared/models/collab_poem_model.dart
/// Collaborative poem model
class CollabPoemModel {
  final String id;
  final String title;
  final String language;
  final String originatorId;
  final String collabType;
  final String status;
  final List<StanzaModel> stanzas;
  final int contributorsCount;
  final List<String> mood;
  final int likesCount;
  final int commentsCount;
  final int readsCount;
  final double trendingScore;
  final DateTime createdAt;
  final DateTime updatedAt;

  const CollabPoemModel({
    required this.id,
    required this.title,
    required this.language,
    required this.originatorId,
    required this.collabType,
    required this.status,
    required this.stanzas,
    required this.contributorsCount,
    required this.mood,
    required this.likesCount,
    required this.commentsCount,
    required this.readsCount,
    required this.trendingScore,
    required this.createdAt,
    required this.updatedAt,
  });

  factory CollabPoemModel.fromJson(Map<String, dynamic> json) {
    final stanzasList = json['stanzas'] as List? ?? [];
    return CollabPoemModel(
      id: json['id'] as String,
      title: json['title'] as String,
      language: json['language'] as String? ?? 'en',
      originatorId: json['originatorId'] as String,
      collabType: json['collabType'] as String? ?? 'open',
      status: json['status'] as String? ?? 'open',
      stanzas: stanzasList
          .map((s) => StanzaModel.fromJson(s as Map<String, dynamic>))
          .toList(),
      contributorsCount: json['contributorsCount'] as int? ?? 0,
      mood: json['mood'] != null ? List<String>.from(json['mood'] as List) : [],
      likesCount: json['likesCount'] as int? ?? 0,
      commentsCount: json['commentsCount'] as int? ?? 0,
      readsCount: json['readsCount'] as int? ?? 0,
      trendingScore: (json['trendingScore'] as num?)?.toDouble() ?? 0.0,
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
    );
  }
}

/// Stanza model
class StanzaModel {
  final String stanzaId;
  final String authorId;
  final String content;
  final int order;
  final bool isApproved;
  final DateTime createdAt;

  const StanzaModel({
    required this.stanzaId,
    required this.authorId,
    required this.content,
    required this.order,
    required this.isApproved,
    required this.createdAt,
  });

  factory StanzaModel.fromJson(Map<String, dynamic> json) {
    return StanzaModel(
      stanzaId: json['stanzaId'] as String,
      authorId: json['authorId'] as String,
      content: json['content'] as String,
      order: json['order'] as int,
      isApproved: json['isApproved'] as bool? ?? true,
      createdAt: DateTime.parse(json['createdAt'] as String),
    );
  }
}

File: verso-api/src/models/CollabPoem.model.ts
import mongoose, { Document, Schema } from 'mongoose';

export interface ICollabPoem extends Document {
  _id: mongoose.Types.ObjectId;
  title: string;
  language: 'en' | 'bn';
  originatorId: mongoose.Types.ObjectId;
  collabType: 'open' | 'invite-only';
  status: 'open' | 'closed';
  stanzas: {
    stanzaId: mongoose.Types.ObjectId;
    authorId: mongoose.Types.ObjectId;
    content: string;
    order: number;
    isApproved: boolean;
    createdAt: Date;
  }[];
  contributorsCount: number;
  mood: string[];
  likesCount: number;
  commentsCount: number;
  readsCount: number;
  trendingScore: number;
  createdAt: Date;
  updatedAt: Date;
}

const StanzaSchema = new Schema({
  stanzaId: { type: Schema.Types.ObjectId, default: () => new mongoose.Types.ObjectId() },
  authorId: { type: Schema.Types.ObjectId, ref: 'User', required: true },
  content: { type: String, required: true, maxlength: 2000 },
  order: { type: Number, required: true },
  isApproved: { type: Boolean, default: true },
  createdAt: { type: Date, default: Date.now },
});

const CollabPoemSchema = new Schema<ICollabPoem>(
  {
    title: { type: String, required: true, trim: true, maxlength: 200 },
    language: { type: String, enum: ['en', 'bn'], default: 'en', index: true },
    originatorId: { type: Schema.Types.ObjectId, ref: 'User', required: true, index: true },
    collabType: { type: String, enum: ['open', 'invite-only'], default: 'open' },
    status: { type: String, enum: ['open', 'closed'], default: 'open', index: true },
    stanzas: [StanzaSchema],
    contributorsCount: { type: Number, default: 0 },
    mood: { type: [String], default: [], index: true },
    likesCount: { type: Number, default: 0 },
    commentsCount: { type: Number, default: 0 },
    readsCount: { type: Number, default: 0 },
    trendingScore: { type: Number, default: 0, index: true },
  },
  { timestamps: true }
);

CollabPoemSchema.index({ trendingScore: -1 });
CollabPoemSchema.index({ status: 1, createdAt: -1 });
CollabPoemSchema.index({ originatorId: 1, createdAt: -1 });

export const CollabPoem = mongoose.model<ICollabPoem>('CollabPoem', CollabPoemSchema);
